File: postgresql-async/src/test/scala/com/github/mauricio/async/db/postgresql/pool/ConnectionPoolSpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.postgresql.pool

import java.util.UUID

import com.github.mauricio.async.db.pool.{ConnectionPool, PoolConfiguration}
import com.github.mauricio.async.db.postgresql.exceptions.GenericDatabaseException
import com.github.mauricio.async.db.postgresql.{PostgreSQLConnection, DatabaseTestHelper}
import org.specs2.mutable.Specification
import scala.concurrent.ExecutionContext.Implicits.global

object ConnectionPoolSpec {
  val Insert = "insert into transaction_test (id) values (?)"
}

class ConnectionPoolSpec extends Specification with DatabaseTestHelper {

  import ConnectionPoolSpec.Insert

  "pool" should {

    "give you a connection when sending statements" in {

      withPool{
        pool =>
          executeQuery(pool, "SELECT 8").rows.get(0)(0) === 8
          Thread.sleep(1000)
          pool.availables.size === 1
      }

    }

    "give you a connection for prepared statements" in {
      withPool{
        pool =>
          executePreparedStatement(pool, "SELECT 8").rows.get(0)(0) === 8
          Thread.sleep(1000)
          pool.availables.size === 1
      }
    }

    "return an empty map when connect is called" in {
      withPool {
        pool =>
          await(pool.connect) === pool
      }
    }

    "runs commands for a transaction in a single connection" in {

      val id = UUID.randomUUID().toString

      withPool {
        pool =>
          val operations = pool.inTransaction {
            connection =>
              connection.sendPreparedStatement(Insert, List(id)).flatMap {
                result =>
                  connection.sendPreparedStatement(Insert, List(id)).map {
                    failure =>
                      List(result, failure)
                  }
              }
          }

          await(operations) must throwA[GenericDatabaseException]

      }

    }

  }

  def withPool[R]( fn : (ConnectionPool[PostgreSQLConnection]) => R ) : R = {

    val pool = new ConnectionPool( new PostgreSQLConnectionFactory(defaultConfiguration), PoolConfiguration.Default )
    try {
      fn(pool)
    } finally {
      await(pool.disconnect)
    }

  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/pool/ConnectionPoolSpec.kt
package com.github.aysnc.sql.db.integration.pool

import com.github.aysnc.sql.db.integration.DatabaseTestHelper
import com.github.aysnc.sql.db.verifyException
import com.github.jasync.sql.db.ConcreteConnection
import com.github.jasync.sql.db.invoke
import com.github.jasync.sql.db.postgresql.exceptions.GenericDatabaseException
import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.flatMapAsync
import com.github.jasync.sql.db.util.mapAsync
import java.util.UUID
import java.util.concurrent.ExecutionException
import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.assertTrue
import org.junit.Test

class ConnectionPoolSpec : DatabaseTestHelper() {
    private val insertQuery = "insert into transaction_test (id) values (?)"

    @Test
    fun `"pool" should "give you a connection when sending statements"`() {
        withPool { pool ->
            assertThat(executeQuery(pool, "SELECT 8").rows[0](0)).isEqualTo(8)
            Thread.sleep(1000)
            assertThat(pool.idleConnectionsCount).isEqualTo(1)
        }
    }

    @Test
    fun `"pool" should "give you a connection for prepared statements"`() {
        withPool { pool ->
            assertThat(executePreparedStatement(pool, "SELECT 8").rows[0](0)).isEqualTo(8)
            Thread.sleep(1000)
            assertThat(pool.idleConnectionsCount).isEqualTo(1)
        }
    }

    @Test
    fun `"pool" should "return an empty map when connect is called"`() {
        withPool { pool ->
            assertThat(awaitFuture(pool.connect())).isEqualTo(pool)
        }
    }

    @Test
    fun `"pool" should "runs commands for a transaction in a single connection" `() {
        val id = UUID.randomUUID().toString()

        withPool { pool ->
            val operations = pool.inTransaction { connection ->
                connection.sendPreparedStatement(insertQuery, listOf(id))
                    .flatMapAsync(ExecutorServiceUtils.CommonPool) { result ->
                        connection.sendPreparedStatement(insertQuery, listOf(id))
                            .mapAsync(ExecutorServiceUtils.CommonPool) { failure ->
                                listOf(result, failure)
                            }
                    }
            }
            verifyException(ExecutionException::class.java, GenericDatabaseException::class.java) {
                awaitFuture(operations)
            }
        }
    }

    @Test
    fun `pool should run queries in use with single connection`() {
        withPool { pool ->
            pool.use { connection ->
                assertTrue(connection is ConcreteConnection)
                connection.sendQuery("SELECT 2")
            }
        }
    }
}
